src/tools/rbd/action/MirrorPool.cc

```diff
@@ -374,6 +376,159 @@ int execute_info(const po::variables_map &vm) {
+
+  std::string config_path;
+  if (vm.count(at::CONFIG_PATH)) {
+    config_path = vm[at::CONFIG_PATH].as<std::string>();
+  }
+
+  librados::Rados rados;
+  librados::IoCtx io_ctx;
+  r = utils::init(pool_name, &rados, &io_ctx);
+  if (r < 0) {
+    return r;
+  }
+
+  std::map<cls::rbd::MirrorImageStatusState, int> states;
+  r = librbd::cls_client::mirror_image_status_get_summary(&io_ctx, &states);
+  if (r < 0) {
+    std::cerr << "rbd: failed to get status summary for mirrored images: "
+	      << cpp_strerror(r) << std::endl;
+    return r;
+  }
+
+  if (formatter != nullptr) {
+    formatter->open_object_section("status");
+  }
+
+  enum Health {Ok = 0, Warning = 1, Error = 2} health = Ok;
+  const char *names[] = {"OK", "WARNING", "ERROR"};
+  int total = 0;
+
+  for (auto &it : states) {
+    auto &state = it.first;
+    if (health < Warning &&
+	(state != cls::rbd::MIRROR_IMAGE_STATUS_STATE_REPLAYING &&
+	 state != cls::rbd::MIRROR_IMAGE_STATUS_STATE_STOPPED)) {
+      health = Warning;
+    }
+    if (health < Error &&
+	state == cls::rbd::MIRROR_IMAGE_STATUS_STATE_ERROR) {
+      health = Error;
+    }
+    total += it.second;
+  }
+
+  if (formatter != nullptr) {
+    formatter->open_object_section("summary");
+    formatter->dump_string("health", names[health]);
+    formatter->open_object_section("states");
+    for (auto &it : states) {
+      std::stringstream name;
+      name << it.first;
+      formatter->dump_int(name.str().c_str(), it.second);
+    }
+    formatter->close_section(); // states
+    formatter->close_section(); // summary
+  } else {
+    std::cout << "health: " << names[health] << std::endl;
+    std::cout << "images: " << total << " total" << std::endl;
+    for (auto &it : states) {
+      std::cout << "    " << it.second << " " << it.first << std::endl;
+    }
+  }
+
+  int ret = 0;
+
+  if (verbose) {
+    if (formatter != nullptr) {
+      formatter->open_array_section("images");
+    }
+
+    cls::rbd::MirrorImageStatus unknown_status(
+      cls::rbd::MIRROR_IMAGE_STATUS_STATE_UNKNOWN, "status not found");
+
+    std::string last_read = "";
+    int max_read = 1024;
+    do {
+      map<std::string, cls::rbd::MirrorImage> mirror_images;
+      map<std::string, cls::rbd::MirrorImageStatus> statuses;
+      r = librbd::cls_client::mirror_image_status_list(&io_ctx, last_read,
+						       max_read, &mirror_images,
+						       &statuses);
+      if (r < 0) {
+	std::cerr << "rbd: failed to list mirrored image directory: "
+		  << cpp_strerror(r) << std::endl;
+	return r;
+      }
+      for (auto it = mirror_images.begin(); it != mirror_images.end(); ++it) {
+	const std::string &image_id = it->first;
+	std::string &global_image_id = it->second.global_image_id;
+	cls::rbd::MirrorImageStatus &status =
+	  statuses.find(image_id) != statuses.end() ?
+	  statuses[image_id] : unknown_status;
+
+	std::string image_name;
+	r = librbd::cls_client::dir_get_name(&io_ctx, RBD_DIRECTORY, image_id,
```

Comment:
I would load the image directory once before the start of the `do` loop so that you don't need to have individual round-trips to the OSD to convert the image id to image name.